Feature/drf parsers and renderers

fanglong-alliance opened the pull request:
to be used after #13's merged

fanglong-alliance committed:
feat: presto drf parsers and renderers port in

levic commented on python-packages/presto_drf/presto_drf/renderers.py:
```diff
@@ -0,0 +1,10 @@
+    def render(self, data, *args, **kwargs):
+        print("===>", data)
+        data = camelize(data)
+        print("-->", data)
```
Leftover debug code

levic commented on python-packages/presto_drf/presto_drf/parsers.py:
```diff
@@ -0,0 +1,65 @@
+        data_and_files.data._mutable = True
+        if request.META.get(
+            "HTTP_X_RADCOMBINEDFORMPOST"
+        ):  # TODO - Confirm this's what we want to call the header? ie, RAD?
```
a value of `"0"` is going to be considered true, but a value of `""` won't. That seems surprising.

davecoates replied:
I'll change it to use strtobool

levic commented on python-packages/presto_drf/presto_drf/parsers.py:
```diff
@@ -0,0 +1,65 @@
+class CamelCaseMultiPartParser(MultiPartParser):
+    """
+    If we have received any files along with other json data it gets converted
+    on frontend into multi-part post with a single key for all JSON data (formData)
```
Why does this happen? Is it not possible to treat it as a vanilla multipart form?

davecoates replied:
No. Normal json data can still be nested so it comes through serialised under one field

levic commented on python-packages/presto_drf/presto_drf/parsers.py:
```diff
@@ -0,0 +1,65 @@
+            new_key = camel_to_underscore(key)
+            if new_key != key:
+                data_and_files.data.pop(key)
+            data_and_files.data[new_key] = underscoreize(value)
```
I'm not understanding something: keys make sense, but why are the values being changed?

davecoates replied:
No idea... can't see any use for this section at all - should be falling back to super I believe ie. only custom thing happens when the header is set

(this came from djrad)

levic replied:
(Assuming this was removed; looks like the history was rebase so I can't actually see the full context of this comment)

levic commented on python-packages/presto_drf/presto_drf/parsers.py:
```diff
@@ -0,0 +1,65 @@
+
+        try:
+            data = stream.read().decode(encoding)
+            return underscoreize(json.loads(data))
```
One of the reasons the alliance utils `underscorize` was written was to be able to differentiate keys that are field names (and should be mangled) from keys that are mappings of a value1 => value2 (so value1 shouldn't be changed).

There doesn't seem to be any way to implement that here, even with subclassing. At minimum the underscorize() invocation should be through a method that you can override if needed

davecoates replied:
Done

davecoates committed:
fix: changes from review

- Support overriding underscoreize
- Rename to CamelCaseMultiPartJSONParser
- Change header to HTTP_X_MULTIPART_JSON and use strtobool to interpret

davecoates committed:
fix: update testbed to use camel case

davecoates commented:
@levic changes made

levic commented on python-packages/presto_drf/presto_drf/tests/test_parsers.py:
```diff
@@ -0,0 +1,92 @@
+            result, {"key_with_numeric_1": {"key_with_numeric2": 1, "key": 2}}
+        )
+
+    def test_camel_case_json_parser_underscorize_ignore(self):
```
Minor since it's in a test case but this uses underscorize whereas everything else uses underscor**e**ize

levic commented:
looks good to merge apart from that one typo
